astronomy/date_time_body.hpp

```diff
@@ -718,26 +761,180 @@ constexpr Time TimeParser::ToTime() const {
+  // after the period don't contribute to the milliseconds.  No rounding is
+  // necessary if there are less than 5 digits after the period.
+  return digit_count <= 2
+             ? 0
+             : digit_count <= 5
+                   ? digit_range(60 * 60 * 24 * digits, 0, 3)
+                   : (digit_range(60 * 60 * 24 * digits,
+                                  digit_count - 4, digit_count) + 5) / 10;
+}
+
+// Operators.
+
+constexpr bool operator==(Date const& left, Date const& right) {
+  return left.year() == right.year() &&
+         left.month() == right.month() &&
+         left.day() == right.day();
+}
+
+constexpr bool operator!=(Date const& left, Date const& right) {
+  return !(left == right);
+}
+
+constexpr bool operator<(Date const& left, Date const& right) {
+  return left.year() < right.year() ||
+         (left.year() == right.year() &&
+          (left.month() < right.month() ||
+           (left.month() == right.month() && left.day() < right.day())));
+}
+
+constexpr bool operator>(Date const& left, Date const& right) {
+  return right < left;
+}
+
+constexpr bool operator<=(Date const& left, Date const& right) {
+  return !(right < left);
+}
+
+constexpr bool operator>=(Date const& left, Date const& right) {
+  return !(left < right);
+}
+
+constexpr Date operator""_Date(char const* str, std::size_t size) {
+  return DateParser::ParseISO(str, size);
+}
+
+constexpr bool operator==(Time const& left, Time const& right) {
+  return left.hour() == right.hour() &&
+         left.minute() == right.minute() &&
+         left.second() == right.second() &&
+         left.millisecond() == right.millisecond();
+}
+
+constexpr bool operator!=(Time const& left, Time const& right) {
+  return !(left == right);
 }
 
 constexpr Time operator""_Time(char const* str, std::size_t size) {
-  return TimeParser::Parse(str, size);
+  return TimeParser::ParseISO(str, size);
 }
 
-// DateTime parsing.
+constexpr bool operator==(DateTime const& left, DateTime const& right) {
+  return left.normalized_end_of_day().date() ==
+             right.normalized_end_of_day().date() &&
+         left.normalized_end_of_day().time() ==
+             right.normalized_end_of_day().time();
+}
+
+constexpr bool operator!=(DateTime const& left, DateTime const& right) {
+  return !(left == right);
+}
 
 constexpr DateTime operator""_DateTime(char const* str, std::size_t size) {
   // Given correctness of the date and time parts of the string, this check
   // ensures that either both are in basic format or both are in extended
   // format.
   return CHECKING(
       contains(str, size, '-') == contains(str, size, ':'),
-      DateTime(
-          operator""_Date(str, index_of(str, size, 'T')),
-          operator""_Time(str + index_of(str, size, 'T') + 1,
-                          size - (index_of(str, size, 'T') + 1))).checked());
+      size >= 2 && str[0] == 'J' && str[1]
+          ? contains(str, size - 1, '.')
```

Comment:
I think this means that the parser doesn't support something like `"MJD1729."`. Is that correct? In any case, please comment to specify precisely what format is allowed, as is done for ISO dates.